wear/src/main/java/io/homeassistant/companion/android/home/MainViewModel.kt

```diff
@@ -173,14 +167,47 @@ class MainViewModel @Inject constructor(
         }
     }
 
+    private fun updateEntityStates(entity: Entity<*>) {
+        if (supportedDomains().contains(entity.domain)) {
+            entities[entity.entityId] = entity
+            // add to cache if part of favorites
+            if (favoriteEntityIds.value.contains(entity.entityId)) {
+                addCachedFavorite(entity.entityId)
+            }
+        }
+    }
+
+    fun updateUI() {
+        if (loadingState.value == LoadingState.READY) {
+            viewModelScope.launch {
+                launch {
+                    entityRegistry = homePresenter.getEntityRegistry()
+                }
+                launch {
+                    deviceRegistry = homePresenter.getDeviceRegistry()
+                }
+                launch {
+                    areaRegistry = homePresenter.getAreaRegistry()?.also {
+                        areas.addAll(it)
+                    }
+                }
+                launch {
+                    homePresenter.getEntities()?.forEach {
+                        updateEntityStates(it)
+                    }
+                    updateEntityDomains()
```

Comment:
This should be done after the registries have been loaded, otherwise if the response to getting registries is received later the UI can still be incorrect or an exception is thrown for concurrent modification and reading.

I think this can be achieved by using `async` + `await` instead of `launch` x4.

Reply:
ok got it, using the same method for loading initial states so it waits.